src/cpp/rtps/writer/StatelessWriter.cpp

```diff
@@ -919,8 +919,8 @@ bool StatelessWriter::matched_reader_add(
 
     locator_selector_.add_entry(new_reader->locator_selector_entry());
 
-    // Datasharing readers handle transient history themselves
-    if (!new_reader->is_datasharing_reader() &&
+    // Local datasharing readers handle transient history themselves
```

Comment:
Looking carefully at the condition, it seems that transient history is managed for
a) All local readers
b) Non-data-sharing readers

This means the ones being special are the remote data-sharing readers

```suggestion
    // Remote datasharing readers handle transient history themselves
```